src/nvim/regexp.c

```diff
@@ -1219,4400 +523,723 @@ char_u *skip_regexp(char_u *startp, int dirc, int magic, char_u **newp)
-    int *flagp
-)
+static void initchr(char_u *str)
 {
-  char_u      *ret;
-  char_u      *br;
-  char_u      *ender;
-  int parno = 0;
-  int flags;
-
-  *flagp = HASWIDTH;            /* Tentatively. */
-
-  if (paren == REG_ZPAREN) {
-    /* Make a ZOPEN node. */
-    if (regnzpar >= NSUBEXP)
-      EMSG_RET_NULL(_("E50: Too many \\z("));
-    parno = regnzpar;
-    regnzpar++;
-    ret = regnode(ZOPEN + parno);
-  } else if (paren == REG_PAREN)    {
-    /* Make a MOPEN node. */
-    if (regnpar >= NSUBEXP)
-      EMSG2_RET_NULL(_("E51: Too many %s("), reg_magic == MAGIC_ALL);
-    parno = regnpar;
-    ++regnpar;
-    ret = regnode(MOPEN + parno);
-  } else if (paren == REG_NPAREN)   {
-    /* Make a NOPEN node. */
-    ret = regnode(NOPEN);
-  } else
-    ret = NULL;
-
-  /* Pick up the branches, linking them together. */
-  br = regbranch(&flags);
-  if (br == NULL)
-    return NULL;
-  if (ret != NULL)
-    regtail(ret, br);           /* [MZ]OPEN -> first. */
-  else
-    ret = br;
-  /* If one of the branches can be zero-width, the whole thing can.
-   * If one of the branches has * at start or matches a line-break, the
-   * whole thing can. */
-  if (!(flags & HASWIDTH))
-    *flagp &= ~HASWIDTH;
-  *flagp |= flags & (SPSTART | HASNL | HASLOOKBH);
-  while (peekchr() == Magic('|')) {
-    skipchr();
-    br = regbranch(&flags);
-    if (br == NULL || reg_toolong)
-      return NULL;
-    regtail(ret, br);           /* BRANCH -> BRANCH. */
-    if (!(flags & HASWIDTH))
-      *flagp &= ~HASWIDTH;
-    *flagp |= flags & (SPSTART | HASNL | HASLOOKBH);
-  }
+  regparse = str;
+  prevchr_len = 0;
+  curchr = prevprevchr = prevchr = nextchr = -1;
+  at_start = true;
+  prev_at_start = false;
+}
 
-  /* Make a closing node, and hook it on the end. */
-  ender = regnode(
-      paren == REG_ZPAREN ? ZCLOSE + parno :
-      paren == REG_PAREN ? MCLOSE + parno :
-      paren == REG_NPAREN ? NCLOSE : END);
-  regtail(ret, ender);
-
-  /* Hook the tails of the branches to the closing node. */
-  for (br = ret; br != NULL; br = regnext(br))
-    regoptail(br, ender);
-
-  /* Check for proper termination. */
-  if (paren != REG_NOPAREN && getchr() != Magic(')')) {
-    if (paren == REG_ZPAREN)
-      EMSG_RET_NULL(_("E52: Unmatched \\z("));
-    else if (paren == REG_NPAREN)
-      EMSG2_RET_NULL(_(e_unmatchedpp), reg_magic == MAGIC_ALL);
-    else
-      EMSG2_RET_NULL(_(e_unmatchedp), reg_magic == MAGIC_ALL);
-  } else if (paren == REG_NOPAREN && peekchr() != NUL) {
-    if (curchr == Magic(')'))
-      EMSG2_RET_NULL(_(e_unmatchedpar), reg_magic == MAGIC_ALL);
-    else
-      EMSG_RET_NULL(_(e_trailing));             /* "Can't happen". */
-    /* NOTREACHED */
-  }
-  // Here we set the flag allowing back references to this set of
-  // parentheses.
-  if (paren == REG_PAREN) {
-    had_endbrace[parno] = true;  // have seen the close paren
+/// Used in a place where no * or \+ can follow.
+static bool re_mult_next(char *what)
```

Comment:
Is there is reason why this function is put here instead of between `match_with_backref()` and `decomp_T` like upstream?

Reply:
I don't remember if I had any specific reason. I may have shifted it during a rebase or something. I'll move it tho.